File: cathode/src/main/java/net/simonvt/cathode/ui/shows/upcoming/UpcomingAdapter.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.shows.upcoming;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import net.simonvt.cathode.R;
import net.simonvt.cathode.common.ui.FragmentsUtils;
import net.simonvt.cathode.common.ui.adapter.HeaderAdapter;
import net.simonvt.cathode.common.widget.OverflowView.OverflowActionListener;
import net.simonvt.cathode.common.widget.RemoteImageView;
import net.simonvt.cathode.common.widget.TimeStamp;
import net.simonvt.cathode.entity.ShowWithEpisode;
import net.simonvt.cathode.images.ImageType;
import net.simonvt.cathode.images.ImageUri;
import net.simonvt.cathode.provider.util.DataHelper;
import net.simonvt.cathode.ui.dialog.CheckInDialog;
import net.simonvt.cathode.ui.dialog.CheckInDialog.Type;
import net.simonvt.cathode.ui.history.AddToHistoryDialog;
import net.simonvt.cathode.widget.CheckInView;

public class UpcomingAdapter extends HeaderAdapter<ShowWithEpisode, RecyclerView.ViewHolder> {

  public interface Callbacks {

    void onEpisodeClicked(long episodeId, String showTitle);

    void onCheckin(long episodeId);

    void onCancelCheckin();
  }

  private static final int TYPE_ITEM = 0;

  private FragmentActivity activity;

  private Callbacks callbacks;

  public UpcomingAdapter(FragmentActivity activity, Callbacks callbacks) {
    super(activity);
    this.activity = activity;
    this.callbacks = callbacks;

    setHasStableIds(true);
  }

  @Override protected int getItemViewType(int headerRes, ShowWithEpisode item) {
    return TYPE_ITEM;
  }

  @Override protected long getItemId(ShowWithEpisode item) {
    return item.getShow().getId();
  }

  @Override protected RecyclerView.ViewHolder onCreateItemHolder(ViewGroup parent, int viewType) {
    View v = LayoutInflater.from(activity).inflate(R.layout.list_row_upcoming, parent, false);
    final ItemViewHolder holder = new ItemViewHolder(v);
    holder.itemView.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View view) {
        final int position = holder.getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) {
          ShowWithEpisode showWithEpisode = getItem(position);
          if (holder.watching) {
            callbacks.onEpisodeClicked(holder.watchingId, showWithEpisode.getShow().getTitle());
          } else {
            callbacks.onEpisodeClicked(showWithEpisode.getEpisode().getId(),
                showWithEpisode.getShow().getTitle());
          }
        }
      }
    });
    return holder;
  }

  @Override protected RecyclerView.ViewHolder onCreateHeaderHolder(ViewGroup parent) {
    View v =
        LayoutInflater.from(activity).inflate(R.layout.list_row_upcoming_header, parent, false);
    return new HeaderViewHolder((TextView) v);
  }

  @Override public void onViewRecycled(RecyclerView.ViewHolder holder) {
    if (holder instanceof ItemViewHolder) {
      ItemViewHolder itemHolder = (ItemViewHolder) holder;
      itemHolder.checkIn.dismiss();
      itemHolder.checkIn.reset();

      itemHolder.watching = false;
      itemHolder.watchingId = null;
    }
  }

  @Override protected void onBindHeader(RecyclerView.ViewHolder holder, int headerRes) {
    HeaderViewHolder vh = (HeaderViewHolder) holder;
    vh.header.setText(headerRes);
  }

  @Override protected void onBindViewHolder(final RecyclerView.ViewHolder holder,
      ShowWithEpisode showWithEpisode, int position) {
    final ItemViewHolder vh = (ItemViewHolder) holder;

    final long showId = showWithEpisode.getShow().getId();
    final boolean watching = showWithEpisode.getShow().getWatching();

    final int airedCount = showWithEpisode.getShow().getAiredCount();
    final int watchedCount = showWithEpisode.getShow().getWatchedCount();

    final long episodeId = showWithEpisode.getEpisode().getId();
    final long episodeFirstAired = showWithEpisode.getEpisode().getFirstAired();
    final int episodeSeasonNumber = showWithEpisode.getEpisode().getSeason();
    final int episodeNumber = showWithEpisode.getEpisode().getEpisode();
    final boolean watched = showWithEpisode.getEpisode().getWatched();
    final String episodeTitle =
        DataHelper.getEpisodeTitle(activity, showWithEpisode.getEpisode().getTitle(),
            episodeSeasonNumber, episodeNumber, watched, true);
    final Long watchingEpisodeId = showWithEpisode.getShow().getWatchingEpisodeId();

    final String showPosterUri = ImageUri.create(ImageUri.ITEM_SHOW, ImageType.POSTER, showId);

    vh.title.setText(showWithEpisode.getShow().getTitle());
    vh.poster.setImage(showPosterUri);

    vh.watching = watching;
    vh.watchingId = watchingEpisodeId;

    if (watching) {
      vh.nextEpisode.setText(R.string.show_watching);
    } else {
      vh.nextEpisode.setText(episodeTitle);
    }

    vh.firstAired.setVisibility(View.VISIBLE);
    vh.firstAired.setTimeInMillis(episodeFirstAired);

    vh.checkIn.setWatching(watching);
    vh.checkIn.setId(showId);
    vh.checkIn.setListener(new OverflowActionListener() {
      @Override public void onPopupShown() {
      }

      @Override public void onPopupDismissed() {
      }

      @Override public void onActionSelected(int action) {
        final int position = holder.getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) {
          switch (action) {
            case R.id.action_checkin_cancel:
              callbacks.onCancelCheckin();
              vh.checkIn.setWatching(false);
              break;

            case R.id.action_checkin:
              if (!CheckInDialog.showDialogIfNecessary(activity, Type.SHOW, episodeTitle,
                  episodeId)) {
                callbacks.onCheckin(episodeId);
                vh.checkIn.setWatching(true);
              }
              break;

            case R.id.action_history_add:
              FragmentsUtils.instantiate(activity.getSupportFragmentManager(),
                  AddToHistoryDialog.class,
                  AddToHistoryDialog.getArgs(AddToHistoryDialog.Type.EPISODE, episodeId,
                      episodeTitle))
                  .show(activity.getSupportFragmentManager(), AddToHistoryDialog.TAG);
              break;
          }
        }
      }
    });
  }

  public static class HeaderViewHolder extends RecyclerView.ViewHolder {

    TextView header;

    public HeaderViewHolder(TextView header) {
      super(header);
      this.header = header;
    }
  }

  public static class ItemViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.title) TextView title;
    @BindView(R.id.nextEpisode) TextView nextEpisode;
    @BindView(R.id.firstAired) TimeStamp firstAired;
    @BindView(R.id.check_in) CheckInView checkIn;
    @BindView(R.id.poster) RemoteImageView poster;

    boolean watching;
    Long watchingId;

    ItemViewHolder(View v) {
      super(v);
      ButterKnife.bind(this, v);
    }
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/shows/upcoming/UpcomingAdapter.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.shows.upcoming

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import net.simonvt.cathode.R.id
import net.simonvt.cathode.R.layout
import net.simonvt.cathode.R.string
import net.simonvt.cathode.common.ui.FragmentsUtils
import net.simonvt.cathode.common.ui.adapter.HeaderAdapter
import net.simonvt.cathode.common.widget.OverflowView.OverflowActionListener
import net.simonvt.cathode.databinding.ListRowUpcomingBinding
import net.simonvt.cathode.entity.ShowWithEpisode
import net.simonvt.cathode.images.ImageType.POSTER
import net.simonvt.cathode.images.ImageUri
import net.simonvt.cathode.provider.util.DataHelper
import net.simonvt.cathode.ui.dialog.CheckInDialog
import net.simonvt.cathode.ui.dialog.CheckInDialog.Type.SHOW
import net.simonvt.cathode.ui.history.AddToHistoryDialog
import net.simonvt.cathode.ui.history.AddToHistoryDialog.Type.EPISODE

class UpcomingAdapter(
  private val activity: FragmentActivity,
  private val callbacks: Callbacks
) : HeaderAdapter<ShowWithEpisode, ViewHolder>(activity) {

  interface Callbacks {
    fun onEpisodeClicked(episodeId: Long, showTitle: String?)
    fun onCheckin(episodeId: Long)
    fun onCancelCheckin()
  }

  override fun getItemViewType(headerRes: Int, item: ShowWithEpisode): Int {
    return TYPE_ITEM
  }

  override fun getItemId(item: ShowWithEpisode): Long {
    return item.show.id
  }

  override fun onCreateItemHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    val binding = ListRowUpcomingBinding.inflate(LayoutInflater.from(activity), parent, false)
    val holder = ItemViewHolder(binding)
    holder.itemView.setOnClickListener {
      val position = holder.adapterPosition
      if (position != RecyclerView.NO_POSITION) {
        val showWithEpisode = getItem(position)
        if (holder.watching) {
          callbacks.onEpisodeClicked(holder.watchingId!!, showWithEpisode!!.show.title)
        } else {
          callbacks.onEpisodeClicked(showWithEpisode!!.episode.id, showWithEpisode.show.title)
        }
      }
    }
    return holder
  }

  override fun onCreateHeaderHolder(parent: ViewGroup): ViewHolder {
    val v = LayoutInflater.from(activity).inflate(layout.list_row_upcoming_header, parent, false)
    return HeaderViewHolder(v as TextView)
  }

  override fun onViewRecycled(holder: ViewHolder) {
    if (holder is ItemViewHolder) {
      holder.checkIn.dismiss()
      holder.checkIn.reset()
      holder.watching = false
      holder.watchingId = null
    }
  }

  override fun onBindHeader(holder: ViewHolder, headerRes: Int) {
    val vh = holder as HeaderViewHolder
    vh.header.setText(headerRes)
  }

  override fun onBindViewHolder(
    holder: ViewHolder,
    showWithEpisode: ShowWithEpisode,
    position: Int
  ) {
    val vh = holder as ItemViewHolder
    val showId = showWithEpisode.show.id
    val watching = showWithEpisode.show.watching
    val episodeId = showWithEpisode.episode.id
    val episodeFirstAired = showWithEpisode.episode.firstAired
    val episodeSeasonNumber = showWithEpisode.episode.season
    val episodeNumber = showWithEpisode.episode.episode
    val watched = showWithEpisode.episode.watched
    val episodeTitle = DataHelper.getEpisodeTitle(
      activity,
      showWithEpisode.episode.title,
      episodeSeasonNumber,
      episodeNumber,
      watched,
      true
    )
    val watchingEpisodeId = showWithEpisode.show.watchingEpisodeId
    val showPosterUri = ImageUri.create(ImageUri.ITEM_SHOW, POSTER, showId)
    vh.title.text = showWithEpisode.show.title
    vh.poster.setImage(showPosterUri)
    vh.watching = watching
    vh.watchingId = watchingEpisodeId
    if (watching) {
      vh.nextEpisode.setText(string.show_watching)
    } else {
      vh.nextEpisode.text = episodeTitle
    }
    vh.firstAired.visibility = View.VISIBLE
    vh.firstAired.setTimeInMillis(episodeFirstAired)
    vh.checkIn.setWatching(watching)
    vh.checkIn.setId(showId)
    vh.checkIn.setListener(object : OverflowActionListener {
      override fun onPopupShown() {}
      override fun onPopupDismissed() {}
      override fun onActionSelected(action: Int) {
        if (holder.getAdapterPosition() != RecyclerView.NO_POSITION) {
          when (action) {
            id.action_checkin_cancel -> {
              callbacks.onCancelCheckin()
              vh.checkIn.setWatching(false)
            }
            id.action_checkin -> if (!CheckInDialog.showDialogIfNecessary(
                activity,
                SHOW,
                episodeTitle,
                episodeId
              )
            ) {
              callbacks.onCheckin(episodeId)
              vh.checkIn.setWatching(true)
            }
            id.action_history_add -> FragmentsUtils.instantiate(
              activity.supportFragmentManager,
              AddToHistoryDialog::class.java,
              AddToHistoryDialog.getArgs(EPISODE, episodeId, episodeTitle)
            ).show(activity.supportFragmentManager, AddToHistoryDialog.TAG)
          }
        }
      }
    })
  }

  class HeaderViewHolder(var header: TextView) : ViewHolder(header)

  class ItemViewHolder(binding: ListRowUpcomingBinding) : ViewHolder(binding.root) {
    var title = binding.title
    var nextEpisode = binding.nextEpisode
    var firstAired = binding.firstAired
    var checkIn = binding.checkIn
    var poster = binding.poster

    var watching = false
    var watchingId: Long? = null
  }

  companion object {
    private const val TYPE_ITEM = 0
  }

  init {
    setHasStableIds(true)
  }
}
